flow/connectors/postgres/ssh_wrapped_pool.go

```diff
@@ -0,0 +1,153 @@
+package connpostgres
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+
+	"github.com/PeerDB-io/peer-flow/connectors/utils"
+	"github.com/PeerDB-io/peer-flow/generated/protos"
+	"github.com/jackc/pgx/v5/pgxpool"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/crypto/ssh"
+)
+
+type SSHWrappedPostgresPool struct {
+	*pgxpool.Pool
+
+	poolConfig *pgxpool.Config
+	sshConfig  *ssh.ClientConfig
+	sshServer  string
+	once       sync.Once
+	sshClient  *ssh.Client
+	localPort  uint16
+	ctx        context.Context
+	cancel     context.CancelFunc
+}
+
+func NewSSHWrappedPostgresPool(
+	ctx context.Context,
+	poolConfig *pgxpool.Config,
+	sshConfig *protos.SSHConfig,
+) (*SSHWrappedPostgresPool, error) {
+	swCtx, cancel := context.WithCancel(ctx)
+
+	var sshServer string
+	var clientConfig *ssh.ClientConfig
+
+	if sshConfig != nil {
+		sshServer = fmt.Sprintf("%s:%d", sshConfig.Host, sshConfig.Port)
+		var err error
+		clientConfig, err = utils.GetSSHClientConfig(
+			sshConfig.User,
+			sshConfig.Password,
+			sshConfig.PrivateKey,
+		)
+		if err != nil {
+			logrus.Error("Failed to get SSH client config: ", err)
+			return nil, err
+		}
+	}
+
+	pool := &SSHWrappedPostgresPool{
+		poolConfig: poolConfig,
+		sshConfig:  clientConfig,
+		sshServer:  sshServer,
+		ctx:        swCtx,
+		cancel:     cancel,
+	}
+
+	err := pool.connect()
+	if err != nil {
+		return nil, err
+	}
+
+	return pool, nil
+}
+
+func (swpp *SSHWrappedPostgresPool) connect() error {
+	var err error
+	swpp.once.Do(func() {
+		err = swpp.setupSSH(swpp.ctx)
+		if err != nil {
+			return
+		}
+		swpp.Pool, err = pgxpool.NewWithConfig(swpp.ctx, swpp.poolConfig)
+	})
+	return err
+}
+
+func (swpp *SSHWrappedPostgresPool) setupSSH(ctx context.Context) error {
+	if swpp.sshConfig == nil {
+		logrus.Info("SSH config is nil, skipping SSH setup")
+		return nil
+	}
+
+	logrus.Info("Setting up SSH connection to ", swpp.sshServer)
+	var err error
+
+	// Establish an SSH connection
+	swpp.sshClient, err = ssh.Dial("tcp", swpp.sshServer, swpp.sshConfig)
+	if err != nil {
+		return err
+	}
+	logrus.Info("SSH connection established")
+	// Automatically pick an available local port
+	localListener, err := net.Listen("tcp", "localhost:0")
```

Comment:
```suggestion
	localListener, err := net.Listen("unix", tmpFile)
```
Did you evaluate unix domain sockets? They have better perf: https://redis.io/docs/management/optimization/benchmarks

> When the server and client benchmark programs run on the same box, both the TCP/IP loopback and unix domain sockets can be used. Depending on the platform, unix domain sockets can achieve around 50% more throughput than the TCP/IP loopback (on Linux for instance). The default behavior of redis-benchmark is to use the TCP/IP loopback.

Reply:
@serprex I reworked this to just modify the `Dial` func, this seems to be what the author recommends as well: https://github.com/jackc/tern/blob/4ca45ec386dd3b7c2b174a16d9a9522bac392b26/ssh_tunnel.go#L21

Reply:
Dial's first parameter can be `"unix"`

https://github.com/jackc/tern/blob/4ca45ec386dd3b7c2b174a16d9a9522bac392b26/ssh_tunnel.go#L47

Reply:
Got it!